avr-hal-generic/src/spi.rs

```diff
@@ -64,220 +66,258 @@ impl Default for Settings {
     }
 }
 
+/// Internal trait for low-level SPI peripherals
+///
+/// This trait defines the common interface for all SPI peripheral variants.  It is used as an
+/// intermediate abstraction ontop of which the [`Spi`] API is built.  **Prefer using the
+/// [`Spi`] API instead of this trait.**
+pub trait SpiOps<H, SCLK, MOSI, MISO, CS> {
+    fn raw_setup(&mut self, settings: &Settings);
+    fn raw_release(&mut self);
+    fn raw_flush(&mut self) -> nb::Result<(), void::Void>;
+    fn raw_read(&self) -> u8;
+    fn raw_write(&mut self, byte: u8);
+}
 
-/// Implement traits for a SPI interface
-#[macro_export]
-macro_rules! impl_spi {
-    (
-        $(#[$spi_attr:meta])*
-        pub struct $Spi:ident {
-            peripheral: $SPI:ty,
-            pins: {
-                sclk: $sclkmod:ident::$SCLK:ident,
-                mosi: $mosimod:ident::$MOSI:ident,
-                miso: $misomod:ident::$MISO:ident,
-                cs: $csmod:ident::$CS:ident,
-            }
-        }
-    ) => {
-        /// First match was without a 'ChipSelectPin' 
-        /// we set it here to a default name and then 
-        /// recusiv expand to the real inplementation 
-        $crate::impl_spi! {
-            pub struct $Spi {
-                peripheral: $SPI,
-                pins: {
-                    sclk: $sclkmod::$SCLK,
-                    mosi: $mosimod::$MOSI,
-                    miso: $misomod::$MISO,
-                    cs: $csmod::$CS,
-                }
-            }
-            pub struct ChipSelectPin;
-        }
-    };
-    
+/// Wrapper for the CS pin
+///
+/// Used to contain the chip-select pin during operation to prevent its mode from being
+/// changed from Output. This is necessary because the SPI state machine would otherwise
+/// reset itself to SPI slave mode immediately. This wrapper can be used just like an
+/// output pin, because it implements all the same traits from embedded-hal.
+pub struct ChipSelectPin<CSPIN: hal::digital::v2::OutputPin>(CSPIN);
```

Comment:
It is good practice not to have trait bounds on the struct, but only on the impls (if possible).  Otherwise this might become a semver-compat hazard.